File: blue-core/src/blue/score/Score.java
/*
 * blue - object composition environment for csound
 * Copyright (C) 2012
 * Steven Yi <stevenyi@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package blue.score;

import blue.CompileData;
import blue.noteProcessor.NoteProcessorChain;
import blue.noteProcessor.NoteProcessorException;
import blue.score.layers.Layer;
import blue.score.layers.LayerGroup;
import blue.score.layers.LayerGroupProviderManager;
import blue.score.layers.ScoreObjectLayer;
import blue.score.tempo.Tempo;
import blue.soundObject.NoteList;
import blue.soundObject.PolyObject;
import blue.soundObject.SoundObject;
import blue.util.ObservableArrayList;
import blue.utility.ScoreUtilities;
import electric.xml.Element;
import electric.xml.Elements;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 *
 * @author stevenyi
 */
public class Score extends ObservableArrayList<LayerGroup<? extends Layer>> {

    Tempo tempo = null;
    TimeState timeState = null;
    private NoteProcessorChain npc; 

    public static final int SPACER = 36;

    public Score() {
        this(true);
    }

    private Score(boolean populate) {
        if (populate) {
            PolyObject pObj = new PolyObject(true);
            add(pObj);
            timeState = new TimeState();
        }
        npc = new NoteProcessorChain();
        tempo = new Tempo();
    }

    public Score(Score score) {
        timeState = new TimeState(score.timeState);
        npc = new NoteProcessorChain(score.npc);
        tempo = new Tempo(score.tempo);

        for(LayerGroup<? extends Layer> lg :score) {
            add(lg.deepCopyLG());    
        }
    }


    public Tempo getTempo() {
        return tempo;
    }

    public void setTempo(Tempo tempo) {
        this.tempo = tempo;
    }

    public TimeState getTimeState() {
        return timeState;
    }

    public void setTimeState(TimeState timeState) {
        this.timeState = timeState;
    }

    public NoteProcessorChain getNoteProcessorChain() {
        return npc;
    }

    public void setNoteProcessorChain(NoteProcessorChain npc) {
        this.npc = npc;
    }

    public Element saveAsXML(Map<Object, String> objRefMap) {
        Element retVal = new Element("score");
        retVal.addElement(tempo.saveAsXML());
        retVal.addElement(timeState.saveAsXML());
        retVal.addElement(npc.saveAsXML());

        for (LayerGroup layerGroup : this) {
            retVal.addElement(layerGroup.saveAsXML(objRefMap));
        }

        return retVal;
    }

    public static Score loadFromXML(Element data, Map<String, Object> objRefMap) throws Exception {
        Score score = new Score(false);

        Elements nodes = data.getElements();

        LayerGroupProviderManager manager = LayerGroupProviderManager.getInstance();

        while (nodes.hasMoreElements()) {
            Element node = nodes.next();
            switch (node.getName()) {
                case "tempo":
                    score.tempo = Tempo.loadFromXML(node);
                    break;
                case "timeState":
                    score.timeState = TimeState.loadFromXML(node);
                    break;
                case "noteProcessorChain":
                    score.npc = NoteProcessorChain.loadFromXML(node);
                    break;
                default:
                    LayerGroup layerGroup = manager.loadFromXML(node, objRefMap);
                    if (layerGroup == null) {
                        throw new RuntimeException(
                                "Unable to load Score LayerGroup of type: " + node.getName());
                    }
                    score.add(layerGroup);
                    if(layerGroup instanceof PolyObject) {
                        ((PolyObject)layerGroup).setTimeBehavior(SoundObject.TIME_BEHAVIOR_NONE);
                    }
                    break;
            }
        }

        if (score.size() == 0) {
            PolyObject pObj = new PolyObject(true);
            score.add(pObj);
        }

        return score;
    }

    public void processOnLoad() {
        for (LayerGroup layerGroup : this) {
            layerGroup.onLoadComplete();
        }
    }

    public NoteList generateForCSD(CompileData compileData, double startTime, double endTime) throws ScoreGenerationException {
        NoteList noteList = new NoteList();

        boolean soloFound = false;

        for (LayerGroup layerGroup : this) {
            soloFound = layerGroup.hasSoloLayers();
            if (soloFound) {
                break;
            }
        }

        for (LayerGroup layerGroup : this) {
            NoteList nl = layerGroup.generateForCSD(compileData, startTime,
                    endTime, soloFound);
            noteList.merge(nl);
        }

        try {
            ScoreUtilities.applyNoteProcessorChain(noteList, this.npc);
        } catch (NoteProcessorException e) {
            throw new ScoreGenerationException(e);
        }

        return noteList;
    }

    public List<LayerGroup> getLayerGroupsForScoreObjects(Collection<? extends ScoreObject> scoreObjects) {
        List<LayerGroup> retVal = new ArrayList<>();

        for (LayerGroup<? extends Layer> layerGroup : this) {
            for (Layer layer : layerGroup) {
                boolean found = false;
                if (layer instanceof ScoreObjectLayer) {
                    ScoreObjectLayer scoreLayer = (ScoreObjectLayer) layer;
                    if (!Collections.disjoint(scoreLayer, scoreObjects)) {
                        retVal.add(layerGroup);
                        found = true;
                    }
                }

                if (found) {
                    break;
                }
            }
        }

        return retVal;
    }

    /* Returns a flat list of all layers in the Score from each LayerGroup */
    public List<Layer> getAllLayers() {
        List<Layer> retVal = new ArrayList<>();

        for (LayerGroup<? extends Layer> layerGroup : this) {
            retVal.addAll(layerGroup);
        }
        return retVal;
    }

    public int getGlobalLayerIndexForY(int y) {
        int runningY = 0;
        int runningIndex = 0;

        for (LayerGroup<? extends Layer> layerGroup : this) {
            for (Layer layer : layerGroup) {
                if (y <= runningY + layer.getLayerHeight()) {
                    return runningIndex;
                }
                runningY += layer.getLayerHeight();
                runningIndex += 1;
            }
            if (y <= runningY + SPACER) {
                return runningIndex;
            }
            runningY += SPACER;
        }

        return runningIndex - 1;
    }

    public Layer getGlobalLayerForY(int y) {
        int runningY = 0;

        for (LayerGroup<? extends Layer> layerGroup : this) {
            for (Layer layer : layerGroup) {
                if (y <= runningY + layer.getLayerHeight()) {
                    return layer;
                }
                runningY += layer.getLayerHeight();
            }
            if (y <= runningY + SPACER) {
                return null;
            }
            runningY += SPACER;
        }

        return null;
    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 79 * hash + Objects.hashCode(this.tempo);
        hash = 79 * hash + Objects.hashCode(this.timeState);
        return hash;
    }


    @Override
    public boolean equals(Object obj) {
        return obj == this;
    }

     
}

File: blue-ext-jython/src/main/release/pythonLib/blue/pmask/score.py
# score.py

# PMask, a Python implementation of CMask
# Copyright (C) 2000 by Maurizio Umberto Puxeddu
# Copyright (C) 2001 by Maurizio Umberto Puxeddu and Hans Mikelson

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import UserList, types
from exception import *
from generator import *

class ScoreEvent:
    def __init__(self, kind, *args):
        self.kind = kind
        self.pfields = []
        self.pfields.extend(args)
        
    def __str__(self):
        string = self.kind
        for pfield in self.pfields:
            if type(pfield) is types.StringType:
                string += ' "' + pfield + '"'
            else:
                string += ' ' + str(pfield)
        return string

    def writeTo(self, file):
        file.write(str(self) + '\n')

class IStatement(ScoreEvent):
    def __init__(self, number, time, duration, *args):
        apply_list = [self, 'i', number, time, duration]
        apply_list.extend(args)
        apply(ScoreEvent.__init__, apply_list)

class FStatement(ScoreEvent):
    def __init__(self, number, time, length, type, *args):
        apply_list = [self, 'f', number, time, length, type]
        apply_list.extend(args)
        apply(ScoreEvent.__init__, apply_list)

class ScoreSection(UserList.UserList):
    def __init__(self, begin, end,
                 number, time, duration, *args):
        UserList.UserList.__init__(self)

        t = begin
        while t < end:
            apply_list = [number, t, evaluateAt(duration, t)]
            for pfield in args:
                apply_list.append(evaluateAt(pfield, t))
            istatement = apply(IStatement, apply_list)
            self.add(istatement)
            t += evaluateAt(time, t)

    def add(self, istatement):
        if not isinstance(istatement, IStatement):
            raise BadArgument
        self.append(istatement)

    def translate(self, time, mode = 'absolute'):
        if mode == 'absolute':
            tmin = None
            for event in self.data:
                if tmin == None or event.pfields[1] < tmin:
                    tmin = event.pfields[1]
            offset = time - tmin
        else:
            offset = time

        for i in self.data:
            i.pfields[1] += offset

    def reverse(self):
        tmin, tmax = self.tmin_tmax()
        for event in self.data:
            event.pfields[1] = tmax - (event.pfields[1] - tmin)

        self.data.reverse()

    def expand(self, factor, mode = 'forward', duration = 'fixed'):
        tmin, tmax = self.tmin_tmax()
        if mode == 'forward':
            zero = tmin
        elif mode == 'backward':
            zero = tmax
        elif mode == 'center':
            zero = (tmin + tmax) / 2.0
        else:
            raise BadArgument

        for event in self.data:
            event.pfields[1] = zero - (zero - event.pfields[1]) * factor

        if duration != 'fixed':
            for event in self.data:
                event.pfields[2] *= factor

    def adjust(self, time):
        for i in self.data:
            offset = evaluateAt(time, i.pfields[1])
            i.pfields[1] += offset

    def tmin_tmax(self):
        tmin = None
        tmax = None
        for event in self.data:
            if tmin == None or event.pfields[1] < tmin:
                tmin = event.pfields[1]
            if tmax == None or event.pfields[1] > tmax:
                tmax = event.pfields[1]
        return tmin, tmax

    def begin_end(self):
        begin = None
        end = None
        for event in self.data:
            if begin == None or event.pfields[1] < begin:
                begin = event.pfields[1]
            if end == None or event.pfields[1] + event.pfields[2] > end:
                end = event.pfields[1] + event.pfields[2] 
        return begin, end
        
    def __str__(self):
        if len(self.data):
            s = str(self.data[0])
            for i in self.data[1:]:
                s += '\n' + str(i)
            return s
        return ''

    def writeTo(self, file):
        s = str(self)
        if s != '':
            file.write(str(self) + '\n')
    
def serializeSections(time, mode = 'joint', *args):
    startTime = time
    for section in args:
        if not isinstance(section, ScoreSection):
            raise BadArgument
        if startTime != None:
            section.translate(startTime)
            
        if mode == 'joint':
            tmin, startTime = section.tmin_tmax()
        elif mode == 'disjoint':
            begin, startTime = section.begin_end()
        else:
            raise BadArgument
